### PR TITLE
Feature - Select view modes per target bundle

### PR DESCRIPTION
**Pull request by Sergey-Orlov**
changes accroding to this conversation https://projects.skilld.cloud/skilld/ski-contrib/issues/74#note_265855

Provide next Override options for selecting view modes

Screenshot_2020-04-09 Manage display Example

Screenshot_2020-04-09 Manage display Example(2)

Screenshot_2020-04-09 Manage display Example(1)

### CODE REVIEW

**Review by Stolexiy — APPROVED** [review]

File: src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php [hunk 1]
```diff
@@ -11,22 +11,24 @@
  * Plugin implementation of the 'ParagraphsViewModeSwitcherFormatter' formatter.
```
Comment (waspper) [inline src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php]: Just needed to remove the paragraphs mentioning.

File: advanced_entity_reference_formatter.info.yml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+name: 'Advanced entity reference formatter'
+type: module
+description: 'Provides advanced formatter for entity reference fields, which allows to select view modes for target bundles or items delta separately.'
```
Comment (waspper) [inline advanced_entity_reference_formatter.info.yml]: This is just a small text suggestion:
"Provides advanced formatter for entity reference fields, allowing to select different view modes per target bundle or  delta item.". For both texts: This one, and text into the "README.md".

File: src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php (hunk 1 above)
Comment (waspper) [inline src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php]: Just needed to remove the paragraphs mentioning.

File: src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php [hunk 3]
```diff
@@ -38,48 +40,93 @@ public static function defaultSettings() {
...
+      '#states' => [
+        'visible' => [
+          ':input[name="fields[' . $field_name . '][settings_edit_form][settings][override]"]' => ['value' => 'bundle'],
+        ],
+      ],
+    ];
+    foreach ($target_bundles as $bundle) {
+      $elements['bundle_view_mode'][$bundle] = [
+        '#type' => 'select',
+        '#options' => $view_modes,
```
Comment (Sergey-Orlov) [inline src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php]: Need to use `getViewModeOptionsByBundle()` method to show only view modes for specific bundle

**Comment by andypost** [comment]
Can you clarify in description why module renamed and is the title valid. Would be great to explain it to reviewers

File: src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php [hunk 4]
```diff
@@ -1,32 +1,34 @@
...
   /**
    * {@inheritdoc}
    */
   public static function defaultSettings() {
     return [
       'view_mode' => 'default',
+      'override' => 'none',
+      'bundle_view_mode' => [],
       'view_mode_delta' => [],
       'link' => FALSE,
```
Comment (piggito) [inline src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php]: `view_mode` and `link` settings come from parent so it might be better to remove them here.

File: README.md [hunk 5]
```diff
@@ -1,2 +1,2 @@
-# entity_reference_display_switcher
-Drupal 8 module provides formatter for entity reference fields, which allows to switch entity reference field view mode
+# Advanced Entity Reference Formatter
+Drupal 8 module that provides advanced formatter for entity reference fields, allowing to select different view modes per target bundle or delta item.
```
Comment (piggito) [inline README.md]: "Advanced" sounds like a broad concept and according to description:
`allowing to select different view modes per target bundle or delta item` it is particulary focused on view mode selection so mb something like "View mode - Entity Reference Formatter" may be more accurate.
This is related to @andypost comment at https://github.com/skilld-labs/advanced_entity_reference_formatter/pull/1#issuecomment-578109971

File: config/schema/advanced_entity_reference_formatter.schema.yml [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+  mapping:
+    override:
+      type: string
+      label: 'Override option'
+    bundle_view_mode:
+      type: sequence
+      label: 'Bundle View mode'
+      sequence:
+        type: string
+    view_mode_delta:
```
Comment (piggito) [inline config/schema/advanced_entity_reference_formatter.schema.yml]: Looks like we need a standard for settings naming cause if we have `view_mode_delta` then why is the other setting `bundle_view_mode` instead of `view_mode_bundle`.
Both are similar options: the prior sets view modes per delta and the last sets view modes per bundle.

Moreover, as mentioned on previous comment, this whole module works around the idea of view mode selection so it looks a bit redundant to constantly repeat `view_mode` in settings parameters.

File: src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php [hunk 7]
```diff
@@ -38,48 +40,95 @@ public static function defaultSettings() {
...
+      ],
+    ];
+    foreach ($target_bundles as $bundle) {
+      // Get view modes for current bundle only.
+      $bundle_view_modes = $this->entityDisplayRepository->getViewModeOptionsByBundle($target_type, $bundle);
+      $elements['bundle_view_mode'][$bundle] = [
+        '#type' => 'select',
+        '#options' => $bundle_view_modes,
+        '#title' => $this->t('View mode for bundle %bundle', ['%bundle' => $bundle]),
+        '#default_value' => $settings['bundle_view_mode'][$bundle],
```
Comment (piggito) [inline src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php]: `$settings['bundle_view_mode]` is an empty array by default so this will throw warnings. Need to validate that `$bundle` key exists.

File: src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php [hunk 8]
```diff
@@ -109,7 +158,7 @@ public static function addMoreSubmit(array $form, FormStateInterface $form_state
    * @param \Drupal\Core\Form\FormStateInterface $form_state
    *   The current state of the form.
    */
-  public static function removeSubmit(array $form, FormStateInterface $form_state) {
+  public static function removeDeltaSubmit(array $form, FormStateInterface $form_state) {
     $items = $form_state->get('delta_items');
     $form_state->set('delta_items', $items - 1);
```
Comment (piggito) [inline src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php]: If `$items = 0` then this will set `delta_items` to `-1` which will lead to unexpected behaviors.
Better validate `$items >= 1` before setting `$items - 1`

File: src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php [hunk 9]
```diff
@@ -140,13 +189,26 @@ public static function ajaxSubmit(array $form, FormStateInterface $form_state) {
...
-        '@mode' => isset($view_modes[$view_mode_delta[$delta]]) ? $view_modes[$view_mode_delta[$delta]] : $view_mode_delta[$view_mode_delta[$delta]],
-      ]);
+
+    $override = $this->getSetting('override');
+    if ($override == 'bundle') {
+      $bundle_view_mode = $this->getSetting('bundle_view_mode') ?: [];
+      foreach ($bundle_view_mode as $bundle => $view_mode) {
+        $summary[] = t('Bundle #@bundle Rendered as @mode', [
+          '@bundle' => $bundle,
+          '@mode' => isset($view_modes[$bundle_view_mode[$bundle]]) ? $view_modes[$bundle_view_mode[$bundle]] : $bundle_view_mode[$bundle],
```
Comment (piggito) [inline src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php]: Replace `$bundle_view_mode[$bundle]` by `$view_mode`

File: src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php [hunk 10]
```diff
@@ -140,13 +189,26 @@ public static function ajaxSubmit(array $form, FormStateInterface $form_state) {
...
+        ]);
+      }
+    }
+    elseif ($override == 'delta') {
+      $view_mode_delta = $this->getSetting('view_mode_delta') ?: [];
+      foreach ($view_mode_delta as $delta => $view_mode) {
+        $summary[] = t('Delta #@delta Rendered as @mode', [
+          '@delta' => $delta,
+          '@mode' => isset($view_modes[$view_mode_delta[$delta]]) ? $view_modes[$view_mode_delta[$delta]] : $view_mode_delta[$view_mode_delta[$delta]],
```
Comment (piggito) [inline src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php]: Replace `$bundle_view_mode[$bundle]` by `$view_mode`

Comment (andypost) [reply src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php]: Also amount of isset() should be minified and why there no fallback in initially prepared array?

File: README.md (hunk 5 above)
Comment (Sergey-Orlov) [reply README.md]: It was renamed according to [this discussion](https://projects.skilld.cloud/skilld/ski-contrib/issues/74#note_266428) with @davidferlay

File: src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php (hunk 10 above)
Comment (Sergey-Orlov) [reply src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php]: This is only summary message, fallback is used in `viewElements()`

File: src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php (hunk 4 above)
Comment (Sergey-Orlov) [reply src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php]: fixed

File: config/schema/advanced_entity_reference_formatter.schema.yml (hunk 6 above)
Comment (Sergey-Orlov) [reply config/schema/advanced_entity_reference_formatter.schema.yml]: updated them to `bundle_modes` and `delta_modes`

File: src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php (hunk 7 above)
Comment (Sergey-Orlov) [reply src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php]: fixed

File: src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php (hunk 8 above)
Comment (Sergey-Orlov) [reply src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php]: fixed

File: src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php (hunk 9 above)
Comment (Sergey-Orlov) [reply src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php]: fixed

File: README.md (hunk 5 above)
Comment (Sergey-Orlov) [reply README.md]: @davidferlay what do you think about Juan's comment?

Comment (davidferlay) [reply README.md]: By using `Advanced` in naming, user might expect to find a bunch of cool new additions, whereas we are adding only 1 thing

the core thing of this module is to make "multiple" something that is usually "unique" > I believe this should be self-evident from module naming : 

I guess it should be something like `Entity Reference Multiple View Modes Formatter` but may be simpler)

Comment (davidferlay) [reply README.md]: > sorlov
> Entity Reference Multiple View Modes Formatter too long i suppose
> 
> dferlay
> could be ok for human name, but machine name should be shorter for sure
> 
> sorlov
> maybe entity_reference_dynamic_display ?
> don't like word Multiple here
> as we don't show multiple view modes for reference
> 
> dferlay
> do we not ?)
> 
> sorlov
> we show 1 view mode for 1 referenced entity)
> from my opinion, multiple view modes -> when you show several view modes for 1 entity
> 
> dferlay
> same entity here for ex, but different bundles
> no?
> 
> sorlov
> you show rendered entity, but exact view_mode is selected based on some conditions
> 
> dferlay
> hm
> entity_reference_dynamic_display
> good for me then



TLDR: let's go with `Entity Reference Dynamic Display` (`entity_reference_dynamic_display`)

File: src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php (hunk 10 above)
Comment (piggito) [reply src/Plugin/Field/FieldFormatter/AdvancedEntityReferenceFormatter.php]: I think fallback is in parent function call so I consider this as resolved

**Review by piggito — APPROVED** [review]
LGTM

File: README.md (hunk 5 above)
Comment (piggito) [reply README.md]: Done
